libsql-wal/src/shared_wal.rs

```diff
@@ -0,0 +1,219 @@
+use std::collections::BTreeMap;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::Instant;
+
+use arc_swap::ArcSwap;
+use crossbeam::deque::Injector;
+use crossbeam::sync::Unparker;
+use libsql_sys::wal::PageHeaders;
+use parking_lot::{Mutex, MutexGuard};
+
+use crate::error::{Error, Result};
+use crate::fs::file::FileExt;
+use crate::fs::FileSystem;
+use crate::name::NamespaceName;
+use crate::registry::WalRegistry;
+use crate::segment::current::CurrentSegment;
+use crate::transaction::{ReadTransaction, Savepoint, Transaction, WriteTransaction};
+
+#[derive(Default)]
+pub struct WalLock {
+    tx_id: Mutex<Option<u64>>,
+    /// When a writer is popped from the write queue, its write transaction may not be reading from the most recent
+    /// snapshot. In this case, we return `SQLITE_BUSY_SNAPHSOT` to the caller. If no reads were performed
+    /// with that transaction before upgrading, then the caller will call us back immediately after re-acquiring
+    /// a read mark.
+    /// Without the reserved slot, the writer would be re-enqueued, a writer before it would be inserted,
+    /// and we'd find ourselves in the initial situation. Instead, we use the reserved slot to bypass the queue when the 
+    /// writer tried to re-acquire the write lock.
+    reserved: Mutex<Option<u64>>,
+    next_tx_id: AtomicU64,
+    waiters: Injector<(Unparker, u64)>,
+}
+
+pub struct SharedWal<FS: FileSystem> {
+    current: ArcSwap<CurrentSegment<FS::File>>,
+    wal_lock: Arc<WalLock>,
+    db_file: FS::File,
+    namespace: NamespaceName,
+    registry: Arc<WalRegistry<FS>>,
+}
+
+impl<FS: FileSystem> SharedWal<FS> {
+    pub fn db_size(&self) -> u32 {
+        self.current.load().db_size()
+    }
+
+    #[tracing::instrument(skip_all)]
+    pub fn begin_read(&self, conn_id: u64) -> ReadTransaction<FS::File> {
+        // FIXME: this is not enough to just increment the counter, we must make sure that the segment
+        // is not sealed. If the segment is sealed, retry with the current segment
+        loop {
+            let current = self.current.load();
+            if current.is_sealed() {
+                continue;
```

Comment:
Maybe it's better to always take whatever log is there, and if the current log is sealed on upgrade, return a snapshot error 🤔

Reply:
> and if the current log is sealed on upgrade, return a snapshot error 🤔

why is this an issue? Assuming there are no new writes. If there are new writes, it will fail anyways.